RFC: Provide equivalence of MPICH_ASYNC_PROGRESS

hominhquan opened the pull request:
This PR is follow-up of #13074.

Commit:
OPAL: re-enable async progress thread

- The SW-based async progress thread has been planned long time ago in
  683efcb, but has never been enabled/implemented since.
- This commit enables the spawn of an async progress thread to execute
  _opal_progress() routine when enabled at both compile time and runtime
  (--enable-progress-threads (default=enabled) and MCA variables
  opal_async_progress or mpi_async_progress).
- Fix minor typo in opal_progress.h doxygen comment

Signed-off-by: Minh Quan Ho <minh-quan.ho@sipearl.com>

janjust commented:
@hominhquan thanks for the PR, do you have any performance data and/or testing data?

bosilca commented on config/opal_configure_options.m4:
```diff
@@ -544,9 +544,22 @@ fi
+# Enable progress threads
+#
+
+AC_MSG_CHECKING([if want asynchronous progress threads])
```
There is no benefit of having this define. Build it in, but only selectively enable it.

hominhquan replied:
default now is enabled

bosilca commented on opal/runtime/opal_progress.c:
```diff
@@ -50,6 +50,38 @@ bool opal_progress_debug = false;
+    /* number of events processed.
+     * This is updated by the async progress thread and only used for
+     * internal tracking purpose */
+    opal_atomic_int64_t nb_events_processed;
```
I don't understand why do you need two atomic variables for the bookkeeping. According to your comment, you expect one to be set by the progress thread and read by the others, and one set by the others. Thus, only one needs to be atomic, the `nb_events_reported`.

Looking at the code it seems you can do everything with a single atomic variable, atomically updated by all threads. The progress thread will add the number of events, while the others will compare_and_swap with 0.

hominhquan replied:
You're right, `nb_events_reported` suffices between the two threads. I've added `nb_events_processed` thinking that we may want to keep track of the total number of processed events, since `nb_events_reported` is read-and-clear-and-forget by the main thread. But this may be useless as you say. I'll remove it.

bosilca commented on opal/runtime/opal_progress.c:
```diff
@@ -119,9 +151,33 @@ static void opal_progress_finalize(void)
+        // end of MPI process, reset nb_events_processed and
+        // not nb_events_reported - which was normally resetted by
+        // application threads
+        opal_atomic_swap_64(&thread_arg.nb_events_processed, 0);
```
No, this is not necessarily the end of the MPI process. It is possible that opal_progress will continue to be called from the main thread, so it is important to let it know there were pending events it did not acknowledge yet.

bosilca commented on opal/runtime/opal_progress.c:
```diff
@@ -155,6 +211,37 @@ int opal_progress_init(void)
+
+    if (opal_async_progress_enabled && !opal_async_progress_thread_spawned) {
+        /* prepare the thread */
+        opal_atomic_swap_64(&thread_arg.nb_events_processed, 0);
```
No need for atomic swap, there is no other thread lurking at them.

bosilca commented on opal/runtime/opal_progress.c:
```diff
@@ -224,13 +311,15 @@ int opal_progress(void)
+     * N = 1024 if opal_async_progress_thread_spawned, otherwise N = 8.
      */
-    if (((num_calls++) & 0x7) == 0) {
+    const uint32_t mod = opal_async_progress_thread_spawned ? 0x3FF : 0x7;
```
that is very low priority for lp callbacks, especially taking into account that you enable yield for the progress thread. What's the rationale ?

hominhquan replied:
I just picked an arbitrary bigger number than the current value (8), there is not really heuristic about 1024. If you think of another value I'll be eager to take it.

bosilca replied:
7 was also arbitrary, but at least that gave our event library the opportunity to do some progress from time to time. From that perspective it looks to me as 1024 is a little to long, let's go for something around 256, and will adapt from there if we notice it takes away too many resources from the communication progress.

That being said, in a progress thread it would make sense to have the low priority queue tested on a time-base interval.

hominhquan replied:
set to 256

bosilca commented on opal/runtime/opal_progress.h:
```diff
@@ -162,9 +162,9 @@ OPAL_DECLSPEC int opal_progress_register(opal_progress_callback_t cb);
 
 /**
- * Deregister previously registered event
+ * unregister previously registered event
```
OMPI community uses deregister everywhere. No need to change.

bosilca commented on opal/runtime/opal_progress.h:
```diff
@@ -195,6 +195,9 @@ static inline bool opal_progress_spin(volatile bool *complete)
     return false;
 }
 
+/* shutdown the async progress thread. Do nothing if disabled */
```
This function should not be visible outside `opal_progress.c`. Please make it static.

hominhquan replied:
This function will be called early in the finalize of `ompi_mpi_finalize()`, which is a solution emerged from #13073 to solve the race during tearing down between the progress thread and the `mca_btl_sm_component` destructor executed by the main thread.

bosilca replied:
My bad, I misread the function and I thought it was the thread progress function.

bosilca reviewed (commented):
A step in the right direction. I still have an issue with the fact that the thread is not bound and cannot be bound, but we can address that later.

devreal commented:
@hominhquan What is the impact on collective operations, both in shared and distributed memory? I imagine there to be more contention...

hppritcha commented on ompi/runtime/ompi_mpi_finalize.c:
```diff
@@ -193,9 +193,10 @@ int ompi_mpi_finalize(void)
-    opal_progress_set_event_flag(OPAL_EVLOOP_ONCE | OPAL_EVLOOP_NONBLOCK);
-#endif
+    /* shutdown async progress thread before tearing down further services */
+    if (opal_async_progress_thread_spawned) {
```
This is okay for now but does leave a hole to plug for the sessions model, but since this is an buy-in option for the application user it should be okay for how.

hppritcha reviewed (dismissed):
There are many other things to consider in terms of performance but in terms of changes to just support the async thread capability this looks okay.  Low risk since its buy in.

hominhquan commented:
> @hominhquan What is the impact on collective operations, both in shared and distributed memory? I imagine there to be more contention...

I only compared with `OSU_Ireduce`. I'll run the whole OSU collective suite this night for a larger view.

bosilca commented on opal/runtime/opal_progress.h:
```diff
@@ -142,7 +142,7 @@ OPAL_DECLSPEC void opal_progress_set_event_poll_rate(int microseconds);
  * Prototype for the a progress function callback.  Progress function
  * callbacks can be registered with opal_progress_register() and
- * deregistered with opal_progress_deregister().  It should be noted
+ * deregistered with opal_progress_unregister().  It should be noted
```
Please change back this one too.

hominhquan replied:
This one fixes the doxygen doc to reflect the effectively used `opal_progress_unregister()` function (line 170). 
Or may you want to rename the function everywhere back to `opal_progress_deregister()` ?

bosilca replied:
all good, my bad.

hominhquan commented:
> > @hominhquan What is the impact on collective operations, both in shared and distributed memory? I imagine there to be more contention...
> 
> I only compared with `OSU_Ireduce`. I'll run the whole OSU collective suite this night for a larger view, but only on a single-node shared-mem config, since I don't have access to any cluster system.

@devreal below is our results on a single-node Grace CPU, in which we see the sweet-spot around 128-256 KB per rank. There was degradation on three operations: `osu_iallgather`, `osu_iallgatherv` and `osu_ialltoall`
![osu_speedup_async_progress](https://github.com/user-attachments/assets/32c3ad68-e2b0-4f23-8dc5-78799b06781b)

devreal commented:
Is that a 3x slowdown for small messages? That would be very concerning. Any idea what would cause that?

hominhquan commented:
> Is that a 3x slowdown for small messages? That would be very concerning. Any idea what would cause that?

It would come from the synchronization overhead between the progress thread (now executing `opal_progress()`) and the main thread (waiting for the job done at each `MPI_Test()` or `MPI_Wait()`). This incompressible cost turns out to be important on small communications (Amdahl's law).

devreal commented:
It's not Amdahl's law if the execution gets slower when you add more resources :) In a perfect world, the thread calling `MPI_Iallreduce` would trigger the progress thread, which then immediately goes ahead and executes the required communications before returning control to the main thread. This looks like many 10s of microseconds synchronization overhead, which is more than I would expect.

hominhquan commented:
As said @bosilca in #13074, (back in 2010-2014), it was hard to get an optimal solution for all message sizes and to all use-cases, and I confirm this conclusion. Spawning a thread introduces many side effects that we can hardly measure their impact. The idea behind this patch is to (re)open the door to further improvement and fine-tuning (core-binding ? time-based yield/progression ? work-stealing ?).

bosilca commented:
These results looks suspicious. OSU doesn't do overlap, it basically posts the non-blocking and wait for it.  

For small messages I could understand 10 to 20 percent performance degradation, but not 3x. And for large messages on an iallgather a 2x increase in performance ? Where is that extra bandwidth coming from ?

hominhquan commented:
Recent version of OSU added, for example `osu_ireduce` which [does overlap](https://github.com/lrbison/osu-micro-benchmarks/blob/main/c/mpi/collective/osu_ireduce.c#L182-L192) and measure the **overall time** of `MPI_Ireduce + dummy_compute(latency_seq) + MPI_Wait`, in which `dummy_compute()` simulates a computation period of `latency_seq = elapsed(MPI_Ireduce + MPI_Wait)`. The goal was to see whether the MPI library manages to overlap 100% the non-blocking collective with the `dummy_compute(...)`. They also did the same with other `osu_iallreduce.c`, `osu_ialltoall` or `osu_ibcast.c` etc..

The results I showed above is on the [`overall_time`](https://github.com/lrbison/osu-micro-benchmarks/blob/main/c/util/osu_util_mpi.c#L747) of those non-blocking operations with the latest OSU benchmark (>= 7.x) https://mvapich.cse.ohio-state.edu/benchmarks/

The speedup is expected from the fact that some tasks (e.g. `Ireduce` or purely data bcast/scatter/gather) are done really in parallel of `dummy_compute()` by the new progress thread, and not at the end of the application thread (`MPI_Wait()`).

Again, I know the limitation of OSU of only using two MPI processes, where resource contention is not stressed much far. They mimic anyway a real well-written overlapped non-blocking schema.

hjelmn commented:
I have been out of the loop for awhile but I thought the idea was to do more targeted progress rather than the hammer that is just looping on opal_progress (which is predictably bad-- we knew this years ago). The concept I remember was signaled sends that could then wake up the remote process that would then make progress. Did that go nowhere?

hominhquan commented:
Ping, please tell me if this PR needs more discussion and has its merit to be merged ?

devreal commented on opal/runtime/opal_params_core.h:
```diff
@@ -59,6 +59,7 @@ OPAL_DECLSPEC extern int opal_initialized;
 OPAL_DECLSPEC extern bool opal_built_with_cuda_support;
 OPAL_DECLSPEC extern bool opal_built_with_rocm_support;
 OPAL_DECLSPEC extern bool opal_built_with_ze_support;
+OPAL_DECLSPEC extern bool opal_async_progress_thread_spawned;
```
This patch removes the `OPAL_ENABLE_PROGRESS_THREADS` pre-processor guard from many places (like https://github.com/open-mpi/ompi/pull/13088/files#diff-b23aacc904fb4f13b003d1c716bf6db21c5c490a319b58a3ab70e8f3a8226885L48). Can we define `opal_async_progress_thread_spawned` to `false` in the header if progress threads are explicitly disabled at compile time? That makes the code cleaner and allows the compiler to optimize away code that is explicitly disabled.

hominhquan replied:
Yes, done in new updated version. Thanks for proposal.

devreal reviewed (commented):
I am in favor of this PR. There is (close to) no harm in merging for users who don't enable the progress thread and I believe it gives us a base for further experimentation. 

There are conflicts that need to be resolved though.

jsquyres commented:
@hominhquan Can you rebase this to the tip of `main` and answer @devreal's pending question?

Thanks!

jsquyres reviewed (commented):
Does anything need to be documented about this?  E.g., how does the user activate this functionality?  What are the benefits / drawbacks?

hominhquan commented:
> Does anything need to be documented about this? E.g., how does the user activate this functionality? What are the benefits / drawbacks?

Right , can a new `docs/launching-apps/progress_thread.rst` entry be a plausible place ? I can see some points : 
- how to enable/disable the feature build at configure and, when enabled, how to activate it at runtime.
- Note about some beneficial usecases, and warn about possible performance degradation in other situations.

jsquyres commented:
> Right , can a new `docs/launching-apps/progress_thread.rst` entry be a plausible place ? I can see some points :
> 
> * how to enable/disable the feature build at configure and, when enabled, how to activate it at runtime.
> * Note about some beneficial usecases, and warn about possible performance degradation in other situations.

Sure, that sounds fine.  And possibly also something under https://docs.open-mpi.org/en/v5.0.x/installing-open-mpi/configure-cli-options/.

I have not read the code, but @tonyskjellum and @bosilca's original points on https://github.com/open-mpi/ompi/issues/13074 are well-noted -- this is *not* a silver bullet of async progress and there are many, many tradeoffs -- potentially even a fairly narrow window for where performance gains will be realized.  That should be clearly called out in the docs.

bosilca commented on opal/runtime/opal_progress.c:
```diff
@@ -254,6 +341,48 @@ int opal_progress(void)
+}
+#endif
+
+int opal_progress(void)
```
I don't think it makes sense to track and return the number of event managed by the progress thread. The reason is that it will mostly be inaccurate if there are multiple threads calling into `opal_progress` to check for progress, and that would lead them to make incorrect assumptions (because they will not see events that were `opal_atomic_swap_64` by another thread).

But this is not really an issue, it only affects a single location `ompi/request/req_test.c`, and the impact will be minimal. I propose you remote all atomic operations around `thread_arg.nb_events_reported`, and make `opal_progress` return 0 in all cases. Then, in `ompi/request/req_test.c` you always `goto recheck_request_status` after the call to `opal_progress`.

hominhquan replied:
I added atomic because the existing semantic of `int opal_progress()` is to return the number of processed events. And yes, the only places where the returned value is used seems to be `req_test.c`. 

If we proceed to what you recommend, I think it is better to change the function to `void opal_progress()` to be consistent to the semantic change ? What do you think ?

devreal replied:
> The reason is that it will mostly be inaccurate if there are multiple threads calling into opal_progress to check for progress, and that would lead them to make incorrect assumptions (because they will not see events that were opal_atomic_swap_64 by another thread).

Don't we have the same problem without the progress thread: two threads may call into `opal_progress` and one may progress some operation while the other doesn't. The second thread won't see that there actually was progress on operations.

hominhquan commented on config/opal_configure_options.m4:
```diff
@@ -544,9 +544,21 @@ fi
+    AC_MSG_RESULT([yes])
+    opal_want_progress_threads=1
+fi
+AC_DEFINE_UNQUOTED([OPAL_ENABLE_PROGRESS_THREADS], [$opal_want_progress_threads],
```
if you are mentioning this @devreal's comment https://github.com/open-mpi/ompi/pull/13088/files/41a1b44d2d371250adfdd8311cdf566b1162e42c#r2054858614 , I believe his idea was not to remove completely `OPAL_ENABLE_PROGRESS_THREADS` from everywhere, but rather to somehow set `opal_async_progress_thread_spawned` to compile-time-known when `OPAL_ENABLE_PROGRESS_THREADS=0`, so that compiler can remove some (not all) portions of code guarded by `if (opal_async_progress_thread_spawned)` (e.g. `req_test.c`)

Commit:
docs: add documentation on async progress thread

Signed-off-by: Minh Quan Ho <minh-quan.ho@sipearl.com>

tonyskjellum commented:
I am definitely in favor of asynchronous progress, maybe I misunderstood
your PR ; putting explicit APIs as was in the MPICH APIs in is different
from making asynchronous progress work without app intervention.

Asynchronous progress often works very well.  We have lots of evidence for
that dating back over 25+ years.   Many apps run faster.

Anthony Skjellum, PhD
***@***.***
Cell: +1-205-807-4968




On Thu, Jun 12, 2025 at 3:02 PM Minh Quan Ho ***@***.***>
wrote:

> *hominhquan* left a comment (open-mpi/ompi#13088)
> <https://github.com/open-mpi/ompi/pull/13088#issuecomment-2966910723>
>
> @tonyskjellum <https://github.com/tonyskjellum> We are definitely aware
> of this *not-always-win* feature, extensively studied and discussed in
> the past. However there are still situation today where it can improve
> performance (single-node shared-mem non-blocking comm., spare (or
> communication-dedicated) cores , see #13088 (comment)
> <https://github.com/open-mpi/ompi/pull/13088#issuecomment-2656213388>).
> The beneficial window is narrow, but it exists.
>
> The progress thread, even enabled at configure, is still not activated at
> runtime (i.e no spawning new thread), user must set an env var
> OPAL_ASYNC_PROGRESS=1 to activate it, like what was done in MPICH.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/open-mpi/ompi/pull/13088#issuecomment-2966910723>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAG7ZCNLVSOMCGHX6IZKTN33DGCBBAVCNFSM6AAAAABW5IXQ5GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSNRWHEYTANZSGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

hominhquan commented:
Documentation added, as suggested by @jsquyres.

bosilca commented:
> I am definitely in favor of asynchronous progress, maybe I misunderstood your PR ; putting explicit APIs as was in the MPICH APIs in is different from making asynchronous progress work without app intervention.
> Asynchronous progress often works very well.  We have lots of evidence for that dating back over 25+ years.   Many apps run faster.

But there is also evidence of the contrary, which is why we should not impose a model on users. Moreover, with the advent of parallel computing and the adoption of OpenMP in many applications/libraries, automatic progress thread might turn out to be detrimental, and should be treated with caution and better left to users discretion.

jsquyres commented on docs/launching-apps/progress_thread.rst:
```diff
@@ -0,0 +1,70 @@
+Asynchronous progress thread
+============================
+
+OpenMPI provides an experimental support of software-based asynchronous
```
"Open MPI" (with a space, without the quotes, obviously).  Same correction is needed for the other places in this PR that the docs mention "OpenMPI".

Sorry, I know this is a nit, but this is the official Open MPI docs and we need to get the name right.

jsquyres commented on docs/launching-apps/progress_thread.rst:
```diff
@@ -0,0 +1,70 @@
+When both variables are set, ``OMPI_ASYNC_PROGRESS`` will be preferred to
+``OPAL_ASYNC_PROGRESS``.
+
+.. warning:: Activating progress thread at runtime may bring more harm
+             than improvement to performance. User should try it with caution.
```
I'd soften this.  Maybe:

```rst
.. warning:: Progress threads are a somewhat complicated issue.  Activating them at 
             run time may improve overlap of communication and computation in 
             your application (particularly those with non-blocking communication) 
             which will improve overall performance.  But there may be unintended
             consequences which may degrade overall application performance.
             Users are advised to experiment and see what works best for their 
             applications.
```

jsquyres commented on docs/launching-apps/progress_thread.rst:
```diff
@@ -0,0 +1,70 @@
+---------
+
+A possible beneficial usecase of software progress thread is *intra-node
+shared-mem non-blocking* communication, running on some high core-count CPUs,
```
shared-memory

jsquyres commented on docs/launching-apps/progress_thread.rst:
```diff
@@ -0,0 +1,70 @@
+A possible beneficial usecase of software progress thread is *intra-node
+shared-mem non-blocking* communication, running on some high core-count CPUs,
+on which application may not use all the available cores, or the CPU has some
+reserved cores dedicated to communication tasks. In such configuration, the
```
In such configurations, ...

jsquyres commented on docs/launching-apps/progress_thread.rst:
```diff
@@ -0,0 +1,70 @@
+reserved cores dedicated to communication tasks. In such configuration, the
+latency of some non-blocking collective operations (e.g. ``MPI_Ireduce()``)
+can be improved thanks to arithmetic operations being performed in the
+background by the progress thread, instead of the main thread which would only
+perform computation lastly at the moment of ``MPI_Wait()``, as currently
+implemented in OpenMPI.
```
...background by the progress thread, instead of deferring the computations to being executed by the main thread during ``MPI_Wait()``.

jsquyres commented on docs/launching-apps/progress_thread.rst:
```diff
@@ -0,0 +1,70 @@
+perform computation lastly at the moment of ``MPI_Wait()``, as currently
+implemented in OpenMPI.
+
+On another hand, on systems where *inter-node communication* was already
+offloaded to an appropriate BTL backend and associated hardware devices,
+enabling the software-based progress thread could be harmful to performance,
+since the additional thread will take up more CPU time for mostly no utility.
```
Alternatively, on systems where inter-node communications are already offloaded to dedicated hardware, enabling the software-based progress threads could degrade performance, since the additional thread will force progress up through the CPU and potentially away from more optimized hardware functionality.

jsquyres commented on docs/installing-open-mpi/configure-cli-options/misc.rst:
```diff
@@ -34,6 +34,18 @@ above categories that can be used with ``configure``:
+  for each MPI process to execute the internal communication progression
+  engine. When enabled at build time, the thread can be activated (default =
+  deactivated) at runtime by setting environment variable
+  ``OMPI_ASYNC_PROGRESS`` or ``OPAL_ASYNC_PROGRESS`` to a non-zero value.
```
Can we make this a proper MCA parameter rather than naked env variables?  MCA parameters are The Open MPI Way -- they can be set in a variety of ways (including via env variables).

hominhquan replied:
I've pushed a new version with MCA variables `opal_async_progress` and its synonym `mpi_async_progress`, add in `opal/runtime/opal_params_core.c`

jsquyres commented:
This PR unfortunately has conflicts; is there any interest in reviving it?

hominhquan commented:
Yes, this PR now has conflict, as the time goes by. I'll try to rebase it in the next few days.
EDIT: Conflicts resolved.

jsquyres commented:
@hominhquan Could I ask you not to use merge commits?  We prefer rebasing in this repo. 

Thank you!

devreal commented:
@hominhquan friendly ping. Could you please rebase so we can merge this PR?

amroakmal commented:
Hi, thanks for this work. I tested this PR (commit 20b1be5e, --mca opal_async_progress 1) on two systems and wanted to share:

1- Performance (LLNL Dane): Tested across 5 applications with different communication patterns -- HPCG, Comb, IMB-ASYNC, Kripke, and PETSc's pipelined CG. Across all of them, at both full-node and reduced-core-count configurations, enabling async progress consistently made things slower, never faster. This held even when leaving physical cores free specifically for the progress thread to use.

2- Crash (LLNL Tuo): On this system, enabling the feature segfaults and runs fail. Confirmed by running same thing but disabling the feature and things worked well.

hominhquan commented:
> @hominhquan friendly ping. Could you please rebase so we can merge this PR?

@devreal @jsquyres I've just repushed a rebased version. Sorry for the delayed reply, I was out of office the last three weeks.

hominhquan commented:
> Hi, thanks for this work. I tested this PR (commit [20b1be5](https://github.com/open-mpi/ompi/commit/20b1be5e81ecfb7d337395b4699c69f111205030), --mca opal_async_progress 1) on two systems and wanted to share:
> 
> 1- Performance (LLNL Dane): Tested across 5 applications with different communication patterns -- HPCG, Comb, IMB-ASYNC, Kripke, and PETSc's pipelined CG. Across all of them, at both full-node and reduced-core-count configurations, enabling async progress consistently made things slower, never faster. This held even when leaving physical cores free specifically for the progress thread to use.
> 
> 2- Crash (LLNL Tuo): On this system, enabling the feature segfaults and runs fail. Confirmed by running same thing but disabling the feature and things worked well.

@amroakmal I'm interested if you can share the command line and logs on the segfault, and if possible, any identification of the problematic spot in `gdb` ?

amroakmal commented:
@hominhquan : 
```
mpirun -np 96 --bind-to core \
  --mca opal_async_progress 1 \
  /usr/WS1/abouelmagd1/benchpark/imb_async_test_tuo/IMB-ASYNC sync_pt2pt async_pt2pt \
  -pt2pt_params topology=halo:ndim=1 \
  -workload_params cpu_calculations=true:cycles_per_10usec=23 \
  > imb_async_progress_full_stdout.txt 2>&1
```

Error:
```
[tuolumne1048:176922:0:176922] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x8)
free(): invalid pointer
[tuolumne1048:176904] *** Process received signal ***
[tuolumne1048:176904] Signal: Aborted (6)
```

Things work when the feature is disabled.

========

Stack-trace:
```
[node1:176922:0:176922] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x8)
[node1:176898:0:176898] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x8)
[node1:176919:0:177126] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x1)
free(): invalid pointer
[node1:176904] *** Process received signal ***
[node1:176904] Signal: Aborted (6)
[node1:176904] Signal code:  (-6)
free(): invalid pointer
[node1:176873] *** Process received signal ***
[node1:176873] Signal: Aborted (6)
[node1:176873] Signal code:  (-6)
[node1:176895:0:177015] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x1)
[node1:176879:0:177010] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x1)
[node1:176884:0:177003] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x1)
[node1:176941:0:177009] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x1)

[node1:176873] [ 0] /lib64/libpthread.so.0(+0x12990)[0x155554406990]
[node1:176873] [ 1] /lib64/libc.so.6(gsignal+0x10f)[0x15555406c5af]
[node1:176873] [ 2] /lib64/libc.so.6(abort+0x127)[0x15555403fee5]
[node1:176873] [ 3] /lib64/libc.so.6(+0x8f6a7)[0x1555540ad6a7]
[node1:176873] [ 4] /lib64/libc.so.6(+0x969ac)[0x1555540b49ac]
[node1:176873] [ 5] /lib64/libc.so.6(+0x98534)[0x1555540b6534]
[node1:176873] [ 6] /path/to/openmpi_build/install/lib/libmpi.so.0(mca_pml_cm_recv_request_completion+0x291)[0x1555551f9a41]
[node1:176873] [ 7] /path/to/openmpi_build/install/lib/libmpi.so.0(+0x20aa06)[0x15555514da06]
[node1:176873] [ 8] /path/to/openmpi_build/install/lib/libmpi.so.0(ompi_mtl_ofi_progress_no_inline+0x18f)[0x155555142daf]
[node1:176873] [ 9] /path/to/openmpi_build/install/lib/libopen-pal.so.0(+0x3807c)[0x1555553e407c]
[node1:176873] [10] /path/to/openmpi_build/install/lib/libopen-pal.so.0(+0x38145)[0x1555553e4145]
[node1:176873] [11] /lib64/libpthread.so.0(+0x81ca)[0x1555543fc1ca]
[node1:176873] [12] /lib64/libc.so.6(clone+0x43)[0x155554057953]
[node1:176873] *** End of error message ***

==== backtrace (tid: 177010) ====
 0  /lib64/libucs.so.0(ucs_handle_error+0x2dc) [0x15555309007c]
 1  /lib64/libucs.so.0(+0x3125c) [0x15555309025c]
 2  /lib64/libucs.so.0(+0x3142a) [0x15555309042a]
 3  /lib64/libpthread.so.0(+0x12990) [0x155554406990]
 4  /path/to/openmpi_build/install/lib/libmpi.so.0(mca_pml_cm_recv_request_completion+0x5c5) [0x1555551f9d75]
 5  /path/to/openmpi_build/install/lib/libmpi.so.0(+0x20aa06) [0x15555514da06]
 6  /path/to/openmpi_build/install/lib/libmpi.so.0(ompi_mtl_ofi_progress_no_inline+0x18f) [0x155555142daf]
 7  /path/to/openmpi_build/install/lib/libopen-pal.so.0(+0x3807c) [0x1555553e407c]
 8  /path/to/openmpi_build/install/lib/libopen-pal.so.0(+0x38145) [0x1555553e4145]
 9  /lib64/libpthread.so.0(+0x81ca) [0x1555543fc1ca]
10  /lib64/libc.so.6(clone+0x43) [0x155554057953]
=================================

==== backtrace (tid: 176922) ====
 0  /lib64/libucs.so.0(ucs_handle_error+0x2dc) [0x15555309007c]
 1  /lib64/libucs.so.0(+0x3125c) [0x15555309025c]
 2  /lib64/libucs.so.0(+0x3142a) [0x15555309042a]
 3  /lib64/libpthread.so.0(+0x12990) [0x155554406990]
 4  /lib64/libfabric.so.1(+0x1260d2) [0x1555536d70d2]
 5  /lib64/libfabric.so.1(+0x11cc96) [0x1555536cdc96]
 6  /path/to/openmpi_build/install/lib/libmpi.so.0(+0x2051d3) [0x1555551481d3]
 7  /path/to/openmpi_build/install/lib/libmpi.so.0(+0x2b44d6) [0x1555551f74d6]
 8  /path/to/openmpi_build/install/lib/libmpi.so.0(ompi_coll_base_barrier_intra_bruck+0x9e) [0x1555550736fe]
 9  /path/to/openmpi_build/install/lib/libmpi.so.0(PMPI_Barrier+0x86) [0x155554ff0846]
10  /path/to/imb_async_test/IMB-ASYNC() [0x44b19a]
11  /path/to/imb_async_test/IMB-ASYNC() [0x43d52f]
12  /path/to/imb_async_test/IMB-ASYNC() [0x4117c7]
13  /lib64/libc.so.6(__libc_start_main+0xe5) [0x155554058865]
14  /path/to/imb_async_test/IMB-ASYNC() [0x40ee8e]
=================================

--------------------------------------------------------------------------
prterun noticed that process rank 22 with PID 176879 on node1 exited on
signal 11 (Segmentation fault).
--------------------------------------------------------------------------
```

Commit:
mtl/ofi: be thread-safe when OPAL async progress thread is active

- Races will happen when the internal OPAL progress thread is active since
  ompi_mtl_ofi_progress() may be called from application and opal progress
  thread

Signed-off-by: Minh Quan Ho <minh-quan.ho@sipearl.com>

hominhquan commented:
@amroakmal Thanks for your feedback. Can you retry if the latest commit [44438e7](https://github.com/open-mpi/ompi/pull/13088/commits/44438e7f9759656eddb2511ab1938d01aadd8f34) solves the problem ?

hominhquan commented:
> @amroakmal Thanks for your feedback. Can you retry if the latest commit [44438e7](https://github.com/open-mpi/ompi/pull/13088/commits/44438e7f9759656eddb2511ab1938d01aadd8f34) solves the problem ?

@amroakmal In case this got lost in the noise, gentle ping to know if the new commit fixed the segfault ?

amroakmal commented:
@hominhquan  : Hi, yes, I'm working on it and will get back to you ASAP once I finish retesting it with the new commit. Thank you and sorry for late response.

amroakmal commented:
@hominhquan : Here's the new stackTrace testing with the new commit (baseline, feature is OFF, finished successfully):

```
(myenv-3.11.5) [xxyyzz]$ grep -av "PMIX ERROR\|knem" "$F" | head -80
--------------------------------------------------------------------------
error. Please check with your system administrator to get the problem

Open MPI will try to fall back on another single-copy mechanism if one
is available.  This may result in lower performance.

  Local host: <localhost>
  Errno:      2 (No such file or directory)
--------------------------------------------------------------------------
|------------------------------------------------------------------|
| component_details:                   |                     pt2pt |
| topology                             |                      halo |
| bidirectional                        |                      true |
| ndim                                 |                         1 |
|------------------------------------------------------------------|
|------------------------------------------------------------------|
| component_details:                   |                  workload |
| cpu_calculations                     |                      true |
| gpu_calculations                     |                     false |
| manual_progress                      |                     false |
| cycles_per_10usec                    |                        23 |
| omit_calc_over_est                   |                     false |
|------------------------------------------------------------------|
[xxyyzz:721328:0:721328] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x8)
==== backtrace (tid: 721328) ====
 0  /lib64/libucs.so.0(ucs_handle_error+0x2dc) [0x15555307207c]
 1  /lib64/libucs.so.0(+0x3125c) [0x15555307225c]
 2  /lib64/libucs.so.0(+0x3142a) [0x15555307242a]
 3  /lib64/libpthread.so.0(+0x12990) [0x155554404990]
 4  /lib64/libfabric.so.1(+0x1377d2) [0x1555536ca7d2]
 5  /lib64/libfabric.so.1(+0x12d6d6) [0x1555536c06d6]
 6  /usr/WS1/user/benchpark/openmpi_test_tuo/install/lib/libmpi.so.0(+0x2067a6) [0x1555551477a6]
 7  /usr/WS1/user/benchpark/openmpi_test_tuo/install/lib/libmpi.so.0(+0x2b5bb6) [0x1555551f6bb6]
 8  /usr/WS1/user/benchpark/openmpi_test_tuo/install/lib/libmpi.so.0(ompi_coll_base_barrier_intra_bruck+0x9e) [0x1555550721fe]
 9  /usr/WS1/user/benchpark/openmpi_test_tuo/install/lib/libmpi.so.0(PMPI_Barrier+0x86) [0x155554fef0d6]
10  /usr/WS1/user/benchpark/imb_async_test_tuo/IMB-ASYNC() [0x44c19a]
11  /usr/WS1/user/benchpark/imb_async_test_tuo/IMB-ASYNC() [0x43e52f]
12  /usr/WS1/user/benchpark/imb_async_test_tuo/IMB-ASYNC() [0x4127c7]
13  /lib64/libc.so.6(__libc_start_main+0xe5) [0x155554056865]
14  /usr/WS1/user/benchpark/imb_async_test_tuo/IMB-ASYNC() [0x40fe8e]
=================================
--------------------------------------------------------------------------

prterun noticed that process rank 53 with PID 721328 on node xxyyzz exited on
signal 11 (Segmentation fault).
--------------------------------------------------------------------------
ASYNC() [0x43e52f]
12  /usr/WS1/user/benchpark/imb_async_test_tuo/IMB-ASYNC() [0x4127c7]
13  /lib64/libc.so.6(__libc_start_main+0xe5) [0x155554056865]
14  /usr/WS1/user/benchpark/imb_async_test_tuo/IMB-ASYNC() [0x40fe8e]
=================================
--------------------------------------------------------------------------

prterun noticed that process rank 91 with PID 1362523 on node xxyyzz exited on
signal 11 (Segmentation fault).
--------------------------------------------------------------------------

```

hominhquan commented:
Thanks @amroakmal, we move forward. I'll keep debugging from here, even I feel like I'm solving a thread-safety problem of another component not related to, but manifested as a consequence of this PR.

Since I don't have access to any similar system like yours, and perhaps you already have reproducer scripts in hand. May I ask you to make a joint effort to debug this ? I'll submit patches, and you test untill it pass ?

amroakmal commented:
@hominhquan : Yes, sure, anytime, but I may not have access to that system for long. Whenever you feel you want us to test plz let me know. Also, I think it would be useful if you try running the same problem that I'm running on one of big systems you have, with same parameters and same problem and same thing, hopefully that could be reproduced. Plz tell me too if you want me to share that.
